### PR TITLE
Fix batch eval mode: allow returning undefined XLATensor

### LINKED ISSUE #913: Batchnorm produce wrong gradient in eval mode
```
import torch
import torch_xla
import torch.nn as nn
import pdb

device = 'xla:0'
# device='cpu'
m_cpu = nn.BatchNorm1d(5)
m = nn.BatchNorm1d(5).to(device)
m_cpu.training = False
m.training = False

input_cpu = torch.rand(4, 5, 3, requires_grad=True)
input = input_cpu.to(device).detach()
input.requires_grad = True

output_cpu = m_cpu(input_cpu)
output = m(input)

output_cpu.sum().backward()
output.sum().backward()

print(input_cpu.grad)
print(input.grad)
```

### PR DESCRIPTION
**Pull request by ailzhang**
fix #913

### CODE REVIEW

File: torch_xla/csrc/aten_xla_bridge.cpp [hunk 1]
```diff
@@ -194,6 +194,9 @@ at::Tensor XlaToAtenTensor(XLATensor xla_tensor,
 }
 
 at::Tensor AtenFromXlaTensor(XLATensor xla_tensor) {
+  if (xla_tensor.is_null()) {
```
Comment (dlibenzi) [inline torch_xla/csrc/aten_xla_bridge.cpp]: Maybe a ternary here still leads to acceptably short statement?

File: test/cpp/test_tensor.cpp [hunk 2]
```diff
@@ -370,6 +370,54 @@ TEST_F(TensorTest, TestAvgPool2DNonSquare) {
   }
 }
 
+TEST_F(TensorTest, TestBatchNorm1D) {
```
Comment (dlibenzi) [inline test/cpp/test_tensor.cpp]: I made a comment in #913 already.
Do we really need this?
Why would a backward plus gradient computation matter in eval mode?
Why would we also have a test in test_tensor.cpp for this (instead of test_aten_xla_tensor.cpp)?

File: torch_xla/csrc/tensor_methods.cpp [hunk 3]
```diff
@@ -1466,7 +1466,8 @@ std::tuple<XLATensor, XLATensor, XLATensor> XLATensor::native_batch_norm(
       input.GetIrValue(), weight_value, bias_value, running_mean_value,
       running_var_value, training, eps);
   XLATensor output = input.CreateFrom(ir::Value(node, 0));
-  XLATensor mean = input.CreateFrom(ir::Value(node, 1));
+  XLATensor mean =
```
Comment (dlibenzi) [inline torch_xla/csrc/tensor_methods.cpp]: We already have an `if (training) {` block in there, so maybe:
```
XLATensor mean;
XLATensor variance_inverse;
if (training) {
  ...
}
```

File: test/cpp/test_tensor.cpp (hunk 2 above)
Comment (ailzhang) [reply test/cpp/test_tensor.cpp]: Hi @dlibenzi , `.eval()` is used to switch between BN layer behavior. Although rare, people sometimes use it in training mode. e.g. https://stackoverflow.com/questions/57323023/pytorch-loss-backward-and-optimizer-step-in-eval-mode-with-batch-norm-laye?rq=1

Another reason I want to fix this is - it broke our assumption that commenting out xla override and code should still work. In this case, I commented out `native_batch_norm_backward` and it was broken because our forward returns non-null `save_mean` but pytorch was expecting an undefined tensor. 

I put it in `test_tensor.cpp` mainly because `at::native_batch_norm` returns a tuple of 3 tensors and the last 2 are the ones I want to check. `torch::batch_norm` tested in `test_aten_xla_tensor.cpp` only returns 1 `torch::Tensor`.

Comment (dlibenzi) [reply test/cpp/test_tensor.cpp]: > Hi @dlibenzi , `.eval()` is used to switch between BN layer behavior. Although rare, people sometimes use it in backward. e.g. https://stackoverflow.com/questions/57323023/pytorch-loss-backward-and-optimizer-step-in-eval-mode-with-batch-norm-laye?rq=1

This is easy, and short fix. But one post on StackOverflow should not be reason to change code otherwise ;)


> Another reason I want to fix this is - it broke our assumption that commenting out xla override and code should still work. In this case, I commented out `native_batch_norm_backward` and it was broken because our forward returns non-null `save_mean` but pytorch was expecting an undefined tensor.

Hmm, if we override FWD and BKW, commenting one override and breaking it should not be cause of changing code.

File: torch_xla/csrc/tensor_methods.cpp [hunk 4]
```diff
@@ -1466,9 +1466,12 @@ std::tuple<XLATensor, XLATensor, XLATensor> XLATensor::native_batch_norm(
       input.GetIrValue(), weight_value, bias_value, running_mean_value,
       running_var_value, training, eps);
   XLATensor output = input.CreateFrom(ir::Value(node, 0));
-  XLATensor mean = input.CreateFrom(ir::Value(node, 1));
+  XLATensor mean, variance_inverse;
   if (training) {
+    mean = input.CreateFrom(ir::Value(node, 1));
```
[the remark this answers is not in the document]
Comment (ailzhang) [reply torch_xla/csrc/tensor_methods.cpp]: `mean` is used in the `running_mean` line ;) so this has to be in the beginning

**Comment by ailzhang** [comment]
> Hmm, if we override FWD and BKW, commenting one override and breaking it should not be cause of changing code.
I feel it's good to keep consistent results with aten counterpart in both FWD and BKW, even when they're used separately. Otherwise it might cause a lot of unnecessary confusions in the future. ;)

File: test/cpp/test_tensor.cpp [hunk 5]
```diff
@@ -370,6 +370,54 @@ TEST_F(TensorTest, TestAvgPool2DNonSquare) {
...
+  at::Tensor running_mean =
+      at::zeros({num_features}, at::TensorOptions(at::kFloat));
+  at::Tensor running_var =
+      at::ones({num_features}, at::TensorOptions(at::kFloat));
+  double momentum = 0.1;
+  double eps = 0.5;
+  at::Tensor undef;
+  for (bool training : {true, false}) {
+    for (bool undef_weight_bias : {true, false}) {
+      auto output = at::native_batch_norm(
```
Comment (dlibenzi) [inline test/cpp/test_tensor.cpp]: About the `auto`. This is not for this PR as I see there is some mess around this file, using both `auto` and `at::Tensor`.
The rule for using `auto` for us is if it is a templated type. So this does not qualify.
Will do a PR for it (or you can if you like).

Comment (ailzhang) [reply test/cpp/test_tensor.cpp]: Btw I use `auto` here mainly to avoid spelling out `std::tuple<at::Tensor, at::Tensor, at::Tensor>`, does this qualify?  :P

Comment (dlibenzi) [reply test/cpp/test_tensor.cpp]: `auto` for `std::tuple<>` qualifies ;)

Comment (ailzhang) [reply test/cpp/test_tensor.cpp]: I see, I'll push a commit to fix other usages in this test.

Comment (ailzhang) [reply test/cpp/test_tensor.cpp]: Oh I see what you meant: this file has more `auto`s in other tests, I can send another PR for it after this.
